docs_src/design/adr/0006-Service-Registry.md

```diff
@@ -0,0 +1,183 @@
+values. In this case the service logged the following errors and exited:
+
+```
+level=ERROR app=device-virtual source=service.go:149 msg="DeviceServicForName failed: Get \"http://localhost:3112/api/v1/deviceservice/name/device-virtual\": dial tcp 127.0.0.1:3112: connect: connection refused"
+level=ERROR app=device-virtual source=init.go:45 msg="Couldn't register to metadata service: Get \"http://localhost:3112/api/v1/deviceservice/name/device-virtual\": dial tcp 127.0.0.1:3112: connect: connection refused\n"
+```
+
+**Note** - in order to run this second test, the easiest way to do so is to remove remove and reinstall the snap vs. manually wiping
+out device-virtual's configuration in Consul. I could have also stopped the service, modified the configuration directly in Consul,
+and restarted the service.
+
+#### Registry Client Interface Usage
+Next the service's usage of the go-mod-registry ```Client``` interface was examined:
+
+```
+type Client interface {
+        // Registers the current service with Registry for discover and health check
+        Register() error
+
+        // Un-registers the current service with Registry for discover and health check
+        Unregister() error
+
+        // Simply checks if Registry is up and running at the configured URL
+        IsAlive() bool
+
+        // Gets the service endpoint information for the target ID from the Registry
+        GetServiceEndpoint(serviceId string) (types.ServiceEndpoint, error)
+
+        // Checks with the Registry if the target service is available, i.e. registered and healthy
+        IsServiceAvailable(serviceId string) (bool, error)
+}
+```
+While the SDK does use the ```IsServiceAvailable``` method on startup to confirm that its dependencies are available, it uses client config
+keys to initialize its various REST API endpoints used to interact with Core Data and Metadata.
+
+**NOTE** - the behavior of device-sdk-c has not yet been verified.
+
+### Core and Support Instead
+The same approach was used for Core and Support services (i.e. reviewing the usage of go-mod-bootstrap's ```Client``` interface), and ironically,
+the SMA seems to be the only service in edgex-go that actually queries the registry for service location:
+
+```
+./internal/system/agent/getconfig/executor.go:		ep, err := e.registryClient.GetServiceEndpoint(serviceName)
+./internal/system/agent/direct/metrics.go:		e, err := m.registryClient.GetServiceEndpoint(serviceName)
+```
+
+### Security Proxy Setup
+The security-proxy-setup service also relies on static service address configuration to configure the server routes for each
+of the services accessible through the API Gateway (aka Kong). Although it uses the same TOML-based client config keys as the
+other services, these configuration values are only ever read from the security-proxy-setup's local configuration.toml file,
+as the security services have never supported using our configuration provider (aka Consul).
```

Comment:
What is the proposal on what security proxy setup should do?

If security proxy setup uses the registry, and services haven't registered by the time that proxy setup runs, what then?

Reply:
Well... I imagine the only way this could be done cleanly would be to support lazy initialization. Let's discuss at tomorrow's WG meeting.